# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1424568944 18000
#      Sat Feb 21 20:35:44 2015 -0500
# Node ID bafcfc5b99af40f82c1c351a1728f1015ab16e9b
# Parent  81078b0e39e8a49114e3a14b60382f61e05bc5f1
if save file name is "-" and nargout is 0, write to stdout

* load-save.cc (Fsave): If save file name is "-" and nargout is 0,
write to stdout.  Update doc string to mention file name of "-".

diff --git a/libinterp/corefcn/load-save.cc b/libinterp/corefcn/load-save.cc
--- a/libinterp/corefcn/load-save.cc
+++ b/libinterp/corefcn/load-save.cc
@@ -1471,23 +1471,24 @@ dump_octave_core (void)
                 }
               else
                 warning ("unable to open '%s' for writing...", fname);
             }
         }
     }
 }
 
-DEFUN (save, args, ,
+DEFUN (save, args, nargout,
        "-*- texinfo -*-\n\
 @deftypefn  {Command} {} save file\n\
 @deftypefnx {Command} {} save options file\n\
 @deftypefnx {Command} {} save options file @var{v1} @var{v2} @dots{}\n\
 @deftypefnx {Command} {} save options file -struct @var{STRUCT} @var{f1} @var{f2} @dots{}\n\
-@deftypefnx {Command} {} {@var{s} =} save @samp{-} @var{v1} @var{v2} @dots{}\n\
+@deftypefnx {Command} {} save @code{\"-\"} @var{v1} @var{v2} @dots{}\n\
+@deftypefnx {Built-in Function} {@var{s} =} save (@code{\"-\"} @var{v1} @var{v2} @dots{})\n\
 Save the named variables @var{v1}, @var{v2}, @dots{}, in the file\n\
 @var{file}.  The special filename @samp{-} may be used to return the\n\
 content of the variables as a string.  If no variable names are listed,\n\
 Octave saves all the variables in the current scope.  Otherwise, full\n\
 variable names or pattern syntax can be used to specify the variables to\n\
 save.  If the @option{-struct} modifier is used, fields @var{f1} @var{f2}\n\
 @dots{} of the scalar structure @var{STRUCT} are saved as if they were\n\
 variables with corresponding names.  Valid options for the @code{save}\n\
@@ -1499,16 +1500,19 @@ If save is invoked using the functional 
 @example\n\
 save (\"-option1\", @dots{}, \"file\", \"v1\", @dots{})\n\
 @end example\n\
 \n\
 @noindent\n\
 then the @var{options}, @var{file}, and variable name arguments\n\
 (@var{v1}, @dots{}) must be specified as character strings.\n\
 \n\
+If called with a filename of @qcode{\"-\"}, write the output to stdout\n\
+if nargout is 0, otherwise return the output in a character string.\n\
+\n\
 @table @code\n\
 @item -append\n\
 Append to the destination instead of overwriting.\n\
 \n\
 @item -ascii\n\
 Save a single matrix in a text file without header or any other information.\n\
 \n\
 @item -binary\n\
@@ -1654,20 +1658,24 @@ the file @file{data} in Octave's binary 
       else
 #endif /* HAVE_HDF5 */
         // don't insert any commands here!  the brace below must go
         // with the "else" above!
         {
           if (append)
             warning ("save: ignoring -append option for output to stdout");
 
-          std::ostringstream output_buf;
-          save_vars (argv, i, argc, output_buf, format,
-                     save_as_floats, true);
-          retval = octave_value (output_buf.str());
+          if (nargout == 0)
+            save_vars (argv, i, argc, std::cout, format, save_as_floats, true);
+          else
+            {
+              std::ostringstream output_buf;
+              save_vars (argv, i, argc, output_buf, format, save_as_floats, true);
+              retval = octave_value (output_buf.str());
+            }
         }
     }
 
   // Guard against things like 'save a*', which are probably mistakes...
 
   else if (i == argc - 1 && glob_pattern_p (argv[i]))
     {
       print_usage ();
